runtime/onert/backend/train/ops/PoolLayer.cc

```diff
@@ -25,7 +31,113 @@ namespace train
 namespace ops
 {
 
-PoolLayer::PoolLayer() : cpu::ops::PoolLayer()
+namespace
+{
+
+cpu::ops::PoolType convert_to_infer(const train::ops::PoolType &pool_type)
+{
+  switch (pool_type)
+  {
+    case train::ops::PoolType::kMax:
+      return cpu::ops::PoolType::kMax;
+    default:
+      throw std::runtime_error("PoolLayer: Unsupported pool type yet");
+  }
+}
+
+class MaxPool2D final : public TrainingKernelRegistry
+{
+private:
+  const ir::Activation _activation;
+  const IPortableTensor *_output;
+  nnfw::cker::PoolParams _op_params;
+
+  std::unique_ptr<Tensor> _act_deriv_output;
+  std::unique_ptr<Tensor> _arg_max_index;
+
+public:
+  MaxPool2D(const uint32_t paddingLeft, const uint32_t, const uint32_t, const uint32_t paddingTop,
+            const uint32_t strideWidth, const uint32_t strideHeight, const uint32_t kernelWidth,
+            const uint32_t kernelHeight, const ir::Activation activation,
+            const IPortableTensor *output)
+    : _activation(activation), _output(output)
+  {
+    {
+      _op_params.stride_height = strideHeight;
+      _op_params.stride_width = strideWidth;
+      _op_params.filter_height = kernelHeight;
+      _op_params.filter_width = kernelWidth;
+      _op_params.padding_values.height = (int8_t)paddingTop;
+      _op_params.padding_values.width = (int8_t)paddingLeft;
+      CalculateActivationRange<float>(activation, &_op_params.float_activation_min,
+                                      &_op_params.float_activation_max);
+    }
+
+    _arg_max_index = std::make_unique<Tensor>(_output->get_info(), _output->layout());
+    _arg_max_index->setBuffer(std::make_shared<basic::Allocator>(_output->total_size()));
+
+    if (activation != ir::Activation::NONE)
+    {
+      _act_deriv_output = std::make_unique<Tensor>(_output->get_info(), _output->layout());
+      _act_deriv_output->setBuffer(std::make_shared<basic::Allocator>(_output->total_size()));
+    }
+  };
+
+  ~MaxPool2D() {}
+
+public:
+  void forward(const IPortableTensor *in, IPortableTensor *out)
+  {
+    assert(in->layout() == ir::Layout::NHWC);
+
+    auto out_shape = getShape(out);
+    auto out_data = getBuffer<float>(out);
+    auto arg_max_index = _arg_max_index.get();
+
+    // maxpool forward
+    nnfw::cker::train::MaxPool2D(_op_params, getShape(in), getBuffer<float>(in), out_shape,
+                                 out_data, getBuffer<int>(arg_max_index));
+
+    // activation forward
+    for (int i = 0; i < out_shape.FlatSize(); ++i)
+    {
+      out_data[i] = nnfw::cker::ActivationFunctionWithMinMax(
+        out_data[i], _op_params.float_activation_min, _op_params.float_activation_max);
+    }
```

Comment:
(Q) Why do you implement this activation function here? The original MaxPool.h kernel code implements this function inside the cker. Could you explain your intention?

Reply:
I hope that `forward` and `backward` functions are matched somehow. 
Since the `backward` explicitly calls gradient-calculation function for `maxpool` and `relu`, I hope that `forward` has a similar look.  

I also thought that if `avgpool` is introduced later, this activation-related part could be exported to a common-thing (to share between `avgpool` and `maxpool`).
To export this common part, I thought it might better to have logic explicit here rather than inside the cker.

Reply:
But I don't have a strong opinion about this part. 
So, If you think it might be better to move inside the `cker`, please let me know. :smile:

Reply:
[Additionally, Also here ](https://github.com/Samsung/ONE/pull/11686/files/4c2d6bbd23ce01ef3c3212bc213045fb5d7a5ab8#r1351950822)

Reply:
I like the similar structure between `forward` and `backward` in same operator. But for me, keeping `forward` consistent to existing inference code as much as possible is more important. Please find more https://github.com/Samsung/ONE/pull/11686/files#r1353573206.

Reply:
You can use `cwiseMin()` and `cwieMax()` of `out_mat` in MaxPool.h of cker like below.
https://github.com/Samsung/ONE/blob/31fc7a8267e38cd0879e85b8e6f4cfd942a010f7/compute/cker/include/cker/eigen/eigen_gemm_eigen.h#L84-L87